### PR TITLE
New lazy decorator

### PR DESCRIPTION
**Pull request by bartvm**
@fvisin @rizar (and I believe @adbrebs touched on this in #384)

This is a stricter version of lazy initialization (spin-off of #528). The lazy decorator is now expected to say which values are needed at which stage of the brick life cycle. I also made laziness non-configurable, because of an important point that @fvisin brought up: It makes code annoying to port, because whether things work or not might depend on someone's configuration. I've also never seen someone explicitly set `Brick.lazy = False`.

```python
@lazy(allocation=['args', 'needed', 'for', 'allocation'],
      initialization=['likewise'])
def __init__(self, always_needed, args, needed, for allocation,
             likewise, allocation_var_with_default=None):
    pass
```

The main advantages:

* Arguments that are needed for brick construction are now actually compulsory again; in the current situation arguments needed in `__init__` would just be replaced with `None` and lead to weird errors.
* No more semi-allocated or semi-initialized bricks with cryptic errors in cases where the configuration wasn't set. We can now check in advance if needed configurations for allocation and initialization are missing.
* Simplification because users don't have to choose between lazy and non-lazy anymore. Bricks are just always lazy.

### CODE REVIEW

**Comment by dwf** [comment]
:+1: Like, very much.

**Comment by bartvm** [comment]
So there was a small bug in there that causes bricks that used `super().__init__` to lose their `allocation_args`, so I fixed that. I also made it "optional" in the sense that you can write a brick without using the `@lazy` decorator, and it shouldn't crash.

Lastly, `Initializable` should technically not have `weights_init` as a compulsory argument. `Initializable` bricks pass their initialization configuration on *if they have it*. I ran into this case with `ConvolutionalSequence`: You can either configure it on the sequence, and pass it on to the components, or you can just set it on the children directly. In the second case, the sequence won't have `weights_init` set, but everything is fully configured anyway...

We should either add `weights_init` to the `@lazy` decorator of the bricks that actually use it in their `initialize` method, or we should split `Initializable` into two: Those that pass on weights configuration on, and those that use weights configuration.

**Comment by fvisin** [comment]
Very nice indeed, thank you Bart! 
I am sorry I couldn't help you updating the @lazy annotations, I didn't understand you needed it done so soon.

I would imagine that most of the bricks will  have `weights_init` as a compulsory argument, but also that there will be other cases similar to `ConvolutionalSequence`. In this case maybe the best solution is to be flexible and split `Initializable` in two. Having a superclass of `Initializable` with less restrictive assumptions doesn't seems like a bad idea to me.

**Comment by bartvm** [comment]
Ugh, second annoying corner case:

`Parallel` requires `input_dims` in order to allocate. Its child `Fork` requires `input_dim` and uses it to set `input_dims` in `push_allocation_config`. Currently this crashes: `Fork` refuses to call `push_allocation_config` because it thinks it needs `input_dims`, but it really doesn't.

One naive way of solving this would be to first call `push_allocation_config`, and *then* check whether everything is configured before calling `allocate`. However, this can still lead to strange errors if `push_allocation_config` tries to iterate over `NoneAllocation` objects for example...

Another naive way would be to not inherit necessary configuration definitions from the parent brick in the class hierarchy, but this makes lazy initialization useless in most cases.

The last solution is simple and my suggestion: If a brick sets configurations required by its parent in `push_allocation_config`, it must set the configuration to `None` explicitly after calling `super`.

**Comment by fvisin** [comment]
Although it is a simple solution, I don't like the idea of having as a rule a "workaround" that has to be read of in the docstring of the lazy decorator or in the documentation. I can easily imagine users failing to create a new brick and also failing to find the _intended way_ of dealing with the situation. 

I like better the idea of having two optional parameters `ignore_inherited_allocation` and `ignore_inherited_initialization` (maybe something less verbose) that specify which inherited initialization/allocation will be handled in a custom way. I am not saying I don't like your idea at all, but I think that the Zen of Python applies here (Explicit is better than implicit).

**Comment by bartvm** [comment]
But also "Simple is better than complex." The Zen of Python is awesome because you can justify anything with it :)

On a more serious note: This is a relatively minor corner case, and there's an argument here which I tend to dislike: "The user won't read the documentation, so let's foresee and handle all possible mistakes the user can make." Although that argument has its merit, in many cases, especially the more advanced ones, I think we should stop taking responsibility for a user not reading the documentation carefully.

If this really does turn into a recurring problem, with users complaining about it, we can always reconsider. For now, let's not overengineer things. The point of this PR was mostly to remove lines, not to write new ones.

**Comment by fvisin** [comment]
Ahahah! You beat me at my own game! What can I say? :)

More seriously, in general I agree that we should take the user accountable if he doesn't read the documentation. In this case though I think that being the parameters optional they wouldn't make the general case more verbose, but it would be clear what to do in the corner case. Importantly, all of this could be achieved with few lines of code. That said, given that you beat me at the Zen of python game, if you still think it's better the other way I won't insist. ;)

**Comment by bartvm** [comment]
Practicaly beats purity ;) So I'd prefer to defer the discussion, maybe make an issue out of it so that we don't forget.

Meanwhile, ready for review by whoever wants to have a closer look!

**Comment by rizar** [comment]
I should say that I like the new interface, and maybe we actually should have something like this.

I am a not fan of forcing they class attributes to have the same names as constructor arguments. This violates encapsulation and is a pretty shady restriction. I would prefer something like that instead:

- the same lazy decorator replaces all not given lazy arguments with instances of `MissingAllocationOption` or `MissingInitializationOption`. Different from the current `AllocationNone` those would have the constructor argument name and the brick object as attributes. Whenever these objects are used in any context an exception like this would be thrown: 

```
AllocationError: missing the "input_dim" argument to the "fork" brick.

You get this error because some of lazy constructor arguments was not given and the corresponding
configuration actions were not taken. Blocks can not allocate parameters without you telling it how to do it!
```

If don't miss anything, that would also solve all the issues you have here. Those `MissingAllocationOption` that were never used will not trigger an error, `ConvolitionalSequence` is happy. Those that will be overridden by the subclass brick ('input_dims`) will not cause any issues as well.

**Comment by bartvm** [comment]
Mm, not a bad idea, but a few concerns:

* We're back to semi-allocated and semi-initialized bricks, because we won't know whether we have the needed configuration until we try and fail.
* Writing these `MissingAllocationOption` objects is messy. You want them to raise an error whatever the code tries to do with them, but that could be accessing attributes, accessing keys, iterating over, testing boolean values, getting length, comparing to another object, used in arithmetic operations, used in NumPy/Theano, etc. You would need to intercept any possible call made to any of these actions to make sure you can raise this sensible error. Is there an easy way to do this? If you miss one, there's a chance again that your code silently does the wrong thing.

I've actually always found the fact that lazy arguments need to be set as attributes later on a feature instead of a restriction. "Either pass `foo` to the constructor, or otherwise set `bar` later on" becomes overly complicated. Lazy construction is already a bit dodgy, I like keeping it simple by really just forcing it to be deferred construction: Either pass `foo` now or set it later. I don't see the link with encapsulation by the way?

I thought about the problem with `Fork` that I ran into a bit more, but I actually don't think the current solution is a "workaround" as much as @fvisin suggested: If a parent brick takes responsibility for configuring a particular setting of one of its children, it needs to signal this in some way. The way to do so is by explicitly setting the configuration to something different from `NoneAllocation`.

**Comment by rizar** [comment]
> We're back to semi-allocated and semi-initialized bricks, because we won't know whether we have the needed configuration until we try and fail.

Here I am not sure I follow. Isn't is the same in the current code of this PR? It seemed to me that you check all the attributes in `allocate`, which is exactly the point where my `MissingOption`'s would start panic.

> Writing these MissingAllocationOption objects is messy.

Indeed that's not super easy. Here is what I came up with:

```
class MissingOption(object):

    def __getattr__(self, name):
        self.panic()

    def panic(self, *args, **kwargs):
        raise Exception("I am missing option")

# The full list can be taken from
# https://docs.python.org/2/reference/datamodel.html#special-method-names
for name in ['__len__', '__add__', '__eq__']:
    setattr(MissingOption, name, MissingOption.panic)

mo = MissingOption()

# All of these work as expected
mo == 1
1 == mo
len(mo)
mo + 1

# Found no way to make this crash.
hasattr(mo, 'a')
```

That is if one has his own checks of allocation parameters validity, those will prevail, which I find tolerable.

> Either pass foo now or set it later. I don't see the link with encapsulation by the way?

What I meant is that it forces the interface (the constructor arguments) and the implementation (the objects attributes) to be the same. In a way the attributes are also a part of interface, so that was not the most successful argument.

However I still find it too restrictive and don't like it. Having two `Initializable` bricks is only one of sacrifices that have to be done to make this work, and it alone looks rather confusing. I will try come up with more concrete examples.

**Comment by bartvm** [comment]
> Here I am not sure I follow. Isn't is the same in the current code of this PR? 

If `MissingOption` appears somewhere at the end of the `_allocate` function, everything before that will be executed. By checking explicitly, the `_allocate` function will never be called if one of the configurations isn't set.

> Having two Initializable bricks is only one of sacrifices that have to be done to make this work

I actually think that this is a bug either way: It doesn't make sense to have `weights_init` as a positional argument if it's in fact optional. It's just that this PR brings this problem more to the foreground.

> However I still find it too restrictive and don't like it.

I believe we understand lazy bricks differently then. This is how I see them: We want to avoid the situation where we need to pass the entire configuration for construction/allocation/initialization in one method call, because we sometimes have complex interdependencies of configuration between bricks. We address this is by saying: "Look, we'll take the arguments from `Brick.allocate` and move them to `Brick.__init__`. You can either pass them there, or you can set them anytime between calling `Brick.__init__` and calling `Brick.allocate`." By that reasoning, it makes perfect sense that attributes need to have exactly the same name as the arguments (has in my mind they've always had to).

File: blocks/bricks/base.py [hunk 1]
```diff
@@ -512,7 +505,7 @@ class Brick(Annotation):
 
     Examples
     --------
-    By default, bricks have lazy initialization enabled.
+    Most bricks have lazy initialization enabled.
```
Comment (fvisin) [inline blocks/bricks/base.py]: Shouldn't this be _all_ bricks now?

**Comment by fvisin** [comment]
I'd like to share an idea. The premise is that it just popped into my mind and it has probably some serious limitation or drawbacks that I am not taking into account. Still, I think it might be worth discussing so....here it is.

We could use the decorator to remove from the object the attributes that have not been set in the constructor (i.e., instead of setting them to `NoneAllocation` and `NoneInitialization`, we remove them). We would then keep two lists of the "suppressed attributes", one for allocation and one for initialization. We can then override __setattr__ and re-add the attribute to the object when the user sets its value (if in the list). This would allow us to call `push_allocation_config`, and _then_ to check if everything has been configured before calling `allocate`. And this check would be as simple as checking if the list is empty. What do you think about it?

**Comment by rizar** [comment]
@bartvm , I will sidestep the complex ideological question for now since I have another example in mind: quite often I do something like this:

```
attention = AttentionMechanism(

```

**Comment by rizar** [comment]
@bartvm , I will sidestep complex ideological questions for now since I have another example in mind: quite often I do something like this:

```python
seq_gen = SequenceGenerator(..., Readout(merge=MLP([None, 500, 500, None])))
```

Normally these `None`'s are to be filled by parents, but once in a while something goes wrong and one gets an exception from `shared_floatx_nan`. Your security checks do not work in such case, whereas the concept of `MissingAllocationOption()` is still applicable:

```python
seq_gen = SequenceGenerator(..., Readout(merge=MLP([
    MissingAllocationOption(), 500, 500, MissingAllocationOption()])))
# This can even remember the stack trace when `MissingAllocationOption` was created.
```

That looks quite bulky, but a syntax sugar to make shorter can be always invented. The main point is that "delayed exception" is a much more flexible paradigm of detecting missing configuration.

File: blocks/bricks/base.py (hunk 1 above)
Comment (bartvm) [inline blocks/bricks/base.py]: Nope, a user can technically write his own brick without the `@lazy` decorator. I think this should be allowed (and I actually edited the code specifically to do so) because it can make for easier development when quickly hacking a brick together e.g. you can just do
```python
StupidBrick(Brick):
    def __init__(self, in, out):
        self.parameters.append(theano.shared(np.random.rand((in, out))))

    @application
    def apply(self, x):
        return x + self.parameters[0]
```

**Comment by bartvm** [comment]
If the user wants to pass a complicated object that raises an error when used, he can do that anyway, this PR isn't stopping the user from doing that, nor does your suggestion solve the problem you point out without requiring some sort of user intervention.

"Delayed exception" is more flexible, but it's also significantly more complicated. My paradigm is: Missing arguments can be set as attributes before calling `allocate`/`initialize`. From the signature of `__init__` and the `@lazy` decorator the user can see *what* they need to set and *when* they need to set it. It's clean and simple. Your solution seems far more complex. If a user doesn't pass a particular argument, it is replaced with a special object, or the user needs to pass this special object manually (introducing more custom concepts the user needs to be familiar with). When and how the user then gets rid of this object needs to be deduced from the code and/or docstring, and could require intimate knowledge of the brick.

**Comment by rizar** [comment]
> If the user wants to pass a complicated object that raises an error when used, he can do that anyway, this PR isn't stopping the user from doing that, nor does your suggestion solve the problem you point out without requiring some sort of user intervention.

Well, this some sort of user intervention is very basic: he just should use the right `None`. And it's not such a complicated object, see my implementation above. Of course this "right None" trick can be used in addition to your checks in `allocate`, but we should we have two ways of doing the same thing when one of the two is anyway required and covers all cases?

> "Delayed exception" is more flexible, but it's also significantly more complicated. 

This is very subjective. I find it very unnatural to bind object attributes and constructor arguments. 

> My paradigm is: Missing arguments can be set as attributes before calling allocate/initialize. From the signature of __init__ and the @lazy decorator the user can see what they need to set and when they need to set it. 

As I said I like your new syntax, but I also propose to use it and provide the user with the same information. What differs is the implementation method and yours has certain limitations, which I find quite important. 

> It's clean and simple. Your solution seems far more complex. If a user doesn't pass a particular argument, it is replaced with a special object, or the user needs to pass this special object manually 

Note, only in the case when your solution is helpless.

> (introducing more custom concepts the user needs to be familiar with).

Your solution introduces a non-standard constraint the user has to be familiar with: attribute names and keyword arguments must be same. Learning about either of these two takes 3 minutes of reading docs.

>  When and how the user then gets rid of this object needs to be deduced from the code and/or docstring, and could require intimate knowledge of the brick.

This is a general issue of lazy initialization though. You never know when a particular missed allocation or initialization setting actually arrives to your brick. 

Yes, one would have to read the `Attributes` section to find out how to configure the brick after allocation, similarly to what we have now. In most cases the section will repeat `Parameters` section, but this is standard Python situation. We are not developing a new DSL and you are the one who protects standard Python practices here, remember #489. 

To sum it up: your solution interferes in Python object life-cycle, whereas what I propose is just a light-weight addition on top of fully lazy bricks that tries to catch the error and explain it to the user in a meaningful way. It can be even implemented without decorator like it was originally proposed in #31.

**Comment by dwf** [comment]
> I find it very unnatural to bind object attributes and constructor arguments.

In general, I agree. But note that @bartvm's solution isn't doing it in general, only for allocation and initialization arguments. This is the way it usually is done now, as far as I can tell: if the user wants to specify weights initialization after construction they set the `weights_init` attribute.

I think I may have a way to make @rizar's proposal more tolerable to you, @bartvm: what if we just checked for these special `MissingWhatevers` in the base class `allocate` and `initialize` methods? That would (as far as I can tell) obviate the need for any real use of them in user code.

**Comment by bartvm** [comment]
> This is the way it usually is done now, as far as I can tell: if the user wants to specify weights initialization after construction they set the weights_init attribute.

That's exactly the behavior I think we should follow. Asking a user to pass `weights_init` *or* set a completely different attribute afterwards is counter-intuitive and requires the user to read through every docstring to see what attributes to set for the arguments they didn't pass.

> what if we just checked for these special `MissingWhatevers` in the base class allocate and initialize methods? 

I'm not sure I understand your proposal. If I understand @rizar's suggestion, he never wants to check for them. We just wait to see if the `_allocate` and `_initialize` method end up using an attribute that somehow contains these objects, and if they do, we hope that they get used in such a way that they raise an error (e.g. doesn't use `hasattr`).

If you are proposing to check whether particular attributes have the value `MissingWhatever` during `allocate` and `initialize` before calling `_allocate`/`_initialize`, then this is exactly what I'm doing. But to know which attribute values to check, you need to have a correspondence between the arguments and the attributes (since you can't just check all attributes, because some of them need to be set for allocation, some for initialization).

**Comment by rizar** [comment]
`MissingWhatever` turned out to be a bad idea: C code from numpy is very hard to fool and all feedback you get when calling `numpy.zeros(whatever)` is `TypeError: an integer is required`. Their check types in a rather tough way [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/conversion_utils.c#L771), and have no desire whatsoever to dig that way.

On the other hand this PR does not do the job as well. Apart from `None`'s in lists that I have already mentioned, consider the case of `biases_init`. It is only required when `has_bias == True`, that is it can not be mentioned in the decorator. That is we have to have a separate sanity check for it in e.g. `Linear` if we don't want to have uninformative error messages like those we have now. 

That being said, I think the only place where reliable security checks can be placed are `_allocate` and `_initialize` methods. And the only one who knows how to do it is the `Brick` author. I think we should not waste our time for partial solution of this problem because of our "lean and mean motto" and to avoid code bloat.

However, I find it very nice that the current `@lazy` decorator (a) separates actually mandatory arguments and those that can be provided later (b)  indicates clearly when is the deadline for an option to be provided. We could have it also insert information which argument of which constructor it was in the `AllocationNone` values it creates, and those could be used to form a nice error message in `_alllocate` and `_initialize`. The `_initialize` routine from `Linear` could look like this:

```
def _initialize(self):
    if self.use_bias:
        if not isinstance(self.biases_init, InitializationScheme):
            complain(self.biases_init)
        W, b = self.params
        self.biases_init.initialize(b, self.rng)
    else:
        W, = self.params
    if not isinstance(self.weights_init, InitializationScheme):
        complain(self.weights_init)
    self.weights_init.initialize(W, self.rng)
```

where `complain` would use information from `self.biases_init` to create a nice error message. One could also explore `self.biases_init` in `pdb` to try to understand what the hell is happening.

I would be happy as well if just get rid of `lazy` decorator altogether. Lack of manpower has been mentioned hundred of times recently, but still I feel we avoid simple solutions at all costs...

**Comment by bartvm** [comment]
I'm not a fan of patching the `__init__` function of each `Brick` automatically, I'd rather keep that explicit by requiring a decorator. The other alternative is for people to write something like `def __init__(self, input_dim=NoneAllocation, output_dim=NoneAllocation)`, but that will look horrendous for bricks like `ConvolutionalLayer`, so I prefer separating that information out into a decorator.

Of course the precise arguments can only be fully checked by the `Brick` itself, but there's a direct parallel here with normal functions:
```python
def foo(a, b=1, c=2):
    if b == c:
        raise ValueError("b and c cannot have the same value")
```
In thise case, the Python function pre-emptively checks whether `a` was provided, because it was compulsory, and calling `foo()` will obviously crash. If we call `foo(1, 1, 1)` it will only crash inside the function.

What my decorator does is simply mirror this precise behaviour: If an argument isn't given (either in `__init__` or later set as an attribute) it crashes pre-emptively. If the actual *value* of the argument is wrong (like `None` in a list) or if several arguments are values that are incompatible, these checks can clearly only be performed inside of the function. This is why I disagree with the statement that I'm coming up with an entire DSL: I'm simply proposing to more strictly mirror Python's default behaviour in our lazy constructor.

**Comment by rizar** [comment]
> I'm not a fan of patching the __init__ function of each Brick automatically

Sorry, where have I proposed it?

**Comment by bartvm** [comment]
You said you would want to get rid of the decorator entirely. In that case there are two options to do lazy initialization that I can see, which I both addressed: Patching `__init__` automatically, or expecting the signature to contain the `NoneAllocation`/`NoneInitialization` defaults.

**Comment by rizar** [comment]
No, what I meant is that we just use `weights_init=None` for all optional parameters in the good old Python way. That is `Linear` would become

```
    def __init__(self, input_dim=None, output_dim=None, **kwargs):
        super(Linear, self).__init__(**kwargs)
        self.input_dim = input_dim
        self.output_dim = output_dim
```

That way we do not reinvent the wheel and do not add any new concepts to Python.
In Python there are mandatory parameters and optional parameters. Now we try to add some allocation parameters, initialization parameters, which are something in between. Should we really do that? If one can skip a parameter when calling a constructor, then it is optional and a default value must be provided in the constructor signature, period. Let's be lean and mean. 

Now that you added the `parent` links debugging missing configuration becomes much easier, one can trace the issue upwards. Also a minimalistic addition to the code above would be to create custom none objects every time that remember where they were created:

```
    def __init__(self, input_dim=NoneA(), output_dim=NoneA(), **kwargs):
        super(Linear, self).__init__(**kwargs)
        self.input_dim = input_dim
        self.output_dim = output_dim
```

**Comment by bartvm** [comment]
> Also a minimalistic addition to the code above would be to create custom none objects every time that remember where they were created:

That's not too useful though, because you can only remember the line in which you were created, and that will be the same line that every other possibly missing configuration was created.

In general, I'm ambivalent about your proposal. I like the fact that it removes the need for custom decorators, but it ruins the signature of the constructor. There is now way to tell from the constructor anymore what needs to be provided and what doesn't. Each docstring will need to minutely detail this, and currently none of the docstrings do this (even though we said we would). This is actually one of the problems this new decorator addresses: Making it immediately clear what needs to be provided when. 

For `Linear` your solution this isn't too bad, but do we really want
```python
class ConvolutionalLayer(Brick):
    def __init__(self, activation, filter_size=None, num_filters=None, pooling_size=None,
                 num_channels=None, conv_step=(1, 1), pooling_step=None,
                 batch_size=None, image_size=None, border_mode='valid',
                 **kwargs):
```
Now it's impossible to tell from the signature that `num_channels` and `pooling_size` are needed, but `pooling_step` and `batch_size` are optional.

> That way we do not reinvent the wheel and do not add any new concepts to Python.

But we are anyway. Keyword arguments in Python are *optional*. In our case they aren't. Both solutions to the problem are equally ugly IMHO:

1. There can be a discrepancy between the signature (which suggests `num_filters` is required) and the functional call (not passing it works), which is what we currently do. This PR proposes making this discrepancy clearer by having a decorator which shows *which* arguments are optional until *when*.

2. Your suggest of getting rid of `@lazy` entirely introduces a discrepancy between the signature and the actual configuration of the brick. The default keyword arguments would suggest that the brick should be configured fully after constructing it, but the documentation needs to be read to see that this isn't true at all: The brick is worthless until a bunch of attributes have been set.

Both approaches mess with the way the objects are normally constructed in Python. The advantage of the `@lazy` decorator has always been that at least the signature makes sense. This new decorator adds even more information to the signature. Your suggestion would make the actual call more sensible, but it removes all information from the code that would show the user what is the necessary configuration.

**Comment by bartvm** [comment]
Rebased this and wondering whether we can merge.

Regardless of what is the optimal solution, I am convinced that this is at least better than what we have right now. The only restriction it introduces is that arguments must correspond directly to attributes, which is a good thing I think, and 99% of bricks did this anyway (only `Fork` didn't do this).

**Comment by dwf** [comment]
This limitation, while indeed restrictive, means there's a standard way to
set such attributes after construction that is discernable from merely
inspecting the call signature. So I'm +1 for merging.

File: blocks/bricks/base.py [hunk 2]
```diff
@@ -768,64 +740,84 @@ def get_dims(self, names):
...
-        return func(self, *args, **kwargs)
-    wraps(func)(init)
-    return init
+    if not allocation:
+        allocation = []
+    if not initialization:
+        initialization = []
+
+    def lazy_wrapper(init):
+        def strict_merge(values):
```
Comment (rizar) [inline blocks/bricks/base.py]: That's not nice to use `strict_merge` here when in the rest of the library so far `dict_union` is used. When you write a replacement and try to use it you will mostly likely only take care about `dict_union` occurences and forget about this one.

**Comment by rizar** [comment]
Okay, let's merge. I have mixed feelings about this PR, but we can give it a try.

File: blocks/bricks/base.py (hunk 2 above)
Comment (bartvm) [inline blocks/bricks/base.py]: Oops, forgot about this one, just put it in there to test.
